Data/ByteString/Base64/Internal.hs

```diff
@@ -205,48 +206,104 @@ decodeWithTable padding decodeFP bs@(PS !fp !o !l) = unsafePerformIO $
-                return $! fromIntegral v :: IO Word32
-              fill !dp !sp !n
-                | sp >= sEnd = finish n
-                | otherwise = {-# SCC "decodeWithTable/fill" #-} do
-                a <- look sp
-                b <- look (sp `plusPtr` 1)
-                c <- look (sp `plusPtr` 2)
-                d <- look (sp `plusPtr` 3)
-                let w = (a `shiftL` 18) .|. (b `shiftL` 12) .|.
-                        (c `shiftL` 6) .|. d
-                if a == done || b == done
-                  then bail $ "invalid padding near offset " ++
-                       show (sp `minusPtr` sptr)
-                  else if a .|. b .|. c .|. d == x
-                  then bail $ "invalid base64 encoding near offset " ++
-                       show (sp `minusPtr` sptr)
-                  else do
-                    poke8 dp $ fromIntegral (w `shiftR` 16)
-                    if c == done
-                      then finish $ n + 1
-                      else do
-                        poke8 (dp `plusPtr` 1) $ fromIntegral (w `shiftR` 8)
-                        if d == done
-                          then finish $! n + 2
-                          else do
-                            poke8 (dp `plusPtr` 2) $ fromIntegral w
-                            fill (dp `plusPtr` 3) (sp `plusPtr` 4) (n+3)
-          withForeignPtr dfp $ \dptr ->
-            fill dptr (sptr `plusPtr` soff) 0
+      withForeignPtr decodeFP $ \ !decptr ->
+        withForeignPtr sfp $ \sptr ->
+        withForeignPtr dfp $ \dptr -> do
+          let !end = sptr `plusPtr` (slen + soff)
+          decodeLoop
+            decptr
+            (plusPtr sptr soff)
+            dptr
+            end
+            dfp
+
+decodeLoop
+    :: Ptr Word8
+    -> Ptr Word8
+    -> Ptr Word8
+    -> Ptr Word8
+    -> ForeignPtr Word8
+    -> IO (Either String ByteString)
+decodeLoop !dtable !sptr !dptr !end !dfp = go dptr sptr 0
+  where
+    err p = return . Left
+      $ "invalid character at offset: "
+      ++ show (p `minusPtr` sptr)
+
+    padErr p =  return . Left
+      $ "invalid padding at offset: "
+      ++ show (p `minusPtr` sptr)
+
+    look :: Ptr Word8 -> IO Word32
+    look !p = do
+      !i <- peekByteOff p 0 :: IO Word8
+      !v <- peekByteOff dtable (fromIntegral i) :: IO Word8
+      return (fromIntegral v)
+
+    go !dst !src !n
+      | plusPtr src 4 >= end = finish dst src n
+      | otherwise = do
+        !a <- look src
+        !b <- look (src `plusPtr` 1)
+        !c <- look (src `plusPtr` 2)
+        !d <- look (src `plusPtr` 3)
+
+        if
+          | a == 0x63 -> padErr src
+          | b == 0x63 -> padErr (plusPtr src 1)
+          | c == 0x63 -> padErr (plusPtr src 2)
+          | d == 0x63 -> padErr (plusPtr src 3)
+          | a == 0xff -> err src
+          | b == 0xff -> err (plusPtr src 1)
+          | c == 0xff -> err (plusPtr src 2)
+          | d == 0xff -> err (plusPtr src 3)
+          | otherwise -> do
```

Comment:
this also improves error reporting by giving a precise offset per byte instead of per chunk